File: jvm/src/main/java/edu/kit/kastel/vads/compiler/parser/Printer.java
package edu.kit.kastel.vads.compiler.parser;

import de.mr_pine.c0ne.parser.ast.AssignmentTree;
import de.mr_pine.c0ne.parser.ast.BinaryOperationTree;
import de.mr_pine.c0ne.parser.ast.BlockTree;
import de.mr_pine.c0ne.parser.ast.IdentExpressionTree;
import de.mr_pine.c0ne.parser.ast.LValueIdentTree;
import de.mr_pine.c0ne.parser.ast.LiteralTree;
import de.mr_pine.c0ne.parser.ast.NameTree;
import de.mr_pine.c0ne.parser.ast.UnaryOperationTree;
import de.mr_pine.c0ne.parser.ast.ReturnTree;
import de.mr_pine.c0ne.parser.ast.Tree;
import de.mr_pine.c0ne.parser.ast.DeclarationTree;
import de.mr_pine.c0ne.parser.ast.FunctionTree;
import de.mr_pine.c0ne.parser.ast.ProgramTree;
import de.mr_pine.c0ne.parser.ast.StatementTree;
import de.mr_pine.c0ne.parser.ast.TypeTree;

/// This is a utility class to help with debugging the parser.
public class Printer {

    private final Tree ast;
    private final StringBuilder builder = new StringBuilder();
    private boolean requiresIndent;
    private int indentDepth;

    public Printer(Tree ast) {
        this.ast = ast;
    }

    public static String print(Tree ast) {
        Printer printer = new Printer(ast);
        printer.printRoot();
        return printer.builder.toString();
    }

    private void printRoot() {
        printTree(this.ast);
    }

    private void printTree(Tree tree) {
        switch (tree) {
            case BlockTree blockTree -> {
                print("{");
                lineBreak();
                this.indentDepth++;
                for (StatementTree statement : blockTree.getStatements()) {
                    printTree(statement);
                }
                this.indentDepth--;
                print("}");
            }
            case FunctionTree functionTree -> {
                printTree(functionTree.returnType);
                space();
                printTree(functionTree.name);
                print("()");
                space();
                printTree(functionTree.body);
            }
            case NameTree nameTree -> print(nameTree.getName().asString());
            case ProgramTree programTree -> {
                for (FunctionTree function : programTree.getTopLevelTrees()) {
                    printTree(function);
                    lineBreak();
                }
            }
            case TypeTree typeTree -> print(typeTree.type().asString());
            case BinaryOperationTree binaryOperationTree -> {
                print("(");
                printTree(binaryOperationTree.getLhs());
                print(")");
                space();
                this.builder.append(binaryOperationTree.getOperatorType());
                space();
                print("(");
                printTree(binaryOperationTree.getRhs());
                print(")");
            }
            case LiteralTree.LiteralIntTree literalIntTree -> this.builder.append(literalIntTree.getValue());
            case UnaryOperationTree unaryOperationTree -> {
                print("-(");
                printTree(unaryOperationTree.getExpression());
                print(")");
            }
            case AssignmentTree assignmentTree -> {
                printTree(assignmentTree.lValue);
                space();
                this.builder.append(assignmentTree.operator);
                space();
                printTree(assignmentTree.expression);
                semicolon();
            }
            case DeclarationTree declarationTree -> {
                printTree(declarationTree.getTypeDeclaration());
                space();
                printTree(declarationTree.getName());
                if (declarationTree.getInitializer() != null) {
                    print(" = ");
                    printTree(declarationTree.getInitializer());
                }
                semicolon();
            }
            case ReturnTree returnTree -> {
                print("return ");
                printTree(returnTree.expression);
                semicolon();
            }
            case LValueIdentTree lValueIdentTree -> printTree(lValueIdentTree.name);
            case IdentExpressionTree identExpressionTree -> printTree(identExpressionTree.getName());
            default -> throw new IllegalStateException("Unexpected value: " + tree);
        }
    }

    private void print(String str) {
        if (this.requiresIndent) {
            this.requiresIndent = false;
            this.builder.append(" ".repeat(4 * this.indentDepth));
        }
        this.builder.append(str);
    }

    private void lineBreak() {
        this.builder.append("\n");
        this.requiresIndent = true;
    }

    private void semicolon() {
        this.builder.append(";");
        lineBreak();
    }

    private void space() {
        this.builder.append(" ");
    }

}

File: jvm/src/main/java/de/mr_pine/c0ne/parser/Printer.kt
package de.mr_pine.c0ne.parser

import de.mr_pine.c0ne.parser.ast.AssignmentTree
import de.mr_pine.c0ne.parser.ast.BinaryOperationTree
import de.mr_pine.c0ne.parser.ast.BlockTree
import de.mr_pine.c0ne.parser.ast.DeclarationTree
import de.mr_pine.c0ne.parser.ast.DeclaredFunctionTree
import de.mr_pine.c0ne.parser.ast.IdentExpressionTree
import de.mr_pine.c0ne.parser.ast.LValueIdentTree
import de.mr_pine.c0ne.parser.ast.LiteralTree
import de.mr_pine.c0ne.parser.ast.NameTree
import de.mr_pine.c0ne.parser.ast.ProgramTree
import de.mr_pine.c0ne.parser.ast.ReturnTree
import de.mr_pine.c0ne.parser.ast.Tree
import de.mr_pine.c0ne.parser.ast.TypeTree
import de.mr_pine.c0ne.parser.ast.UnaryOperationTree


/** This is a utility class to help with debugging the parser. */
class Printer(private val ast: Tree) {
    private val builder = StringBuilder()
    private var requiresIndent = false
    private var indentDepth = 0

    private fun printRoot() {
        printTree(this.ast)
    }

    private fun printTree(tree: Tree) {
        when (tree) {
            is BlockTree -> {
                print("{")
                lineBreak()
                this.indentDepth++
                for (statement in tree.statements) {
                    printTree(statement)
                }
                this.indentDepth--
                print("}")
            }

            is DeclaredFunctionTree -> {
                printTree(tree.returnTypeTree)
                space()
                printTree(tree.name)
                print("(")
                printTree(tree.parameters)
                print(")")
                space()
                printTree(tree.body)
            }

            is NameTree -> print(tree.name.asString())
            is ProgramTree -> {
                for (function in tree.topLevelTrees) {
                    printTree(function)
                    lineBreak()
                }
            }

            is TypeTree -> print(tree.type.asString())
            is BinaryOperationTree -> {
                print("(")
                printTree(tree.lhs)
                print(")")
                space()
                this.builder.append(tree.operatorType)
                space()
                print("(")
                printTree(tree.rhs)
                print(")")
            }

            is LiteralTree.LiteralIntTree -> this.builder.append(tree.value)
            is UnaryOperationTree -> {
                this.builder.append(tree.operator.type)
                print("(")
                printTree(tree.expression)
                print(")")
            }

            is AssignmentTree -> {
                printTree(tree.lValue)
                space()
                this.builder.append(tree.operator)
                space()
                printTree(tree.expression)
                semicolon()
            }

            is DeclarationTree -> {
                printTree(tree.typeDeclaration)
                space()
                printTree(tree.name)
                if (tree.initializer != null) {
                    print(" = ")
                    printTree(tree.initializer)
                }
                semicolon()
            }

            is ReturnTree -> {
                print("return ")
                printTree(tree.expression)
                semicolon()
            }

            is LValueIdentTree -> printTree(tree.name)
            is IdentExpressionTree -> printTree(tree.name)
            else -> throw IllegalStateException("Unexpected value: $tree")
        }
    }

    private fun print(str: String?) {
        if (this.requiresIndent) {
            this.requiresIndent = false
            this.builder.append(" ".repeat(4 * this.indentDepth))
        }
        this.builder.append(str)
    }

    private fun lineBreak() {
        this.builder.append("\n")
        this.requiresIndent = true
    }

    private fun semicolon() {
        this.builder.append(";")
        lineBreak()
    }

    private fun space() {
        this.builder.append(" ")
    }

    companion object {
        fun print(ast: Tree): String {
            val printer = Printer(ast)
            printer.printRoot()
            return printer.builder.toString()
        }
    }
}
